firedrake/assemble.py

```diff
@@ -1180,25 +1180,31 @@ def allocate(self):
 
     def _apply_bc(self, tensor, bc):
         # TODO Maybe this could be a singledispatchmethod?
-        if isinstance(bc, DirichletBC):
-            self._apply_dirichlet_bc(tensor, bc)
-        elif isinstance(bc, EquationBCSplit):
-            bc.zero(tensor)
-            type(self)(bc.f, bcs=bc.bcs, form_compiler_parameters=self._form_compiler_params, needs_zeroing=False,
-                       zero_bc_nodes=self._zero_bc_nodes, diagonal=self._diagonal).assemble(tensor=tensor)
-        else:
-            raise AssertionError
-
-    def _apply_dirichlet_bc(self, tensor, bc):
-        if not self._zero_bc_nodes:
+        # Handle special diagonal case first.
+        if self._diagonal:
+            if not isinstance(bc, DirichletBC):
+                raise TypeError(f"diagonal expects a DirichletBC: got {bc}")
+            # Ignore self._zero_bc_nodes.
             tensor_func = tensor.riesz_representation(riesz_map="l2")
-            if self._diagonal:
-                bc.set(tensor_func, 1)
-            else:
-                bc.apply(tensor_func)
+            bc.set(tensor_func, 1)
             tensor.assign(tensor_func.riesz_representation(riesz_map="l2"))
```

Comment:
I don't think bc.set does type checking, so the riesz_representation could go away. This will cause a conflict, as I also modified this code here https://github.com/firedrakeproject/firedrake/pull/3922